feat(msi): add fileAssociation support for MSI target

aplum opened the pull request:
Also fix iconId sometimes containing invalid characters, and iconId config option being ignored (was added in #6247, but it doesn't seem to have been used).

Marking this as a draft and targeting [v23.0.0-alpha](https://github.com/electron-userland/electron-builder/tree/v23.0.0-alpha) because I think there's some more related changes that should be made:
- I'm not sure why the iconId config option was added, and I think it should be removed. I don't see why anyone would want to manually configure it when we should be able to generate it ourselves. Even if the product name contains all invalid-for-MSI-Id characters, we could still generate a valid unique Id somehow (base64, seeded RNG, whatever).
  - Even if we _do_ want to keep the option to manually configure it, it's now too specific – we also need a product-specific Id prefix for ProgId Ids with this PR, and possibly others in the future.
- I didn't support FileAssociation.icon yet, but I'd be willing to if someone wants to lend a hand. The WiX side should be easy (add more [Icon](https://github.com/aplum/electron-builder/blob/4090ab62f16f05c5aa2bf7b2491415efd4594158/packages/app-builder-lib/templates/msi/template.xml#L23) elements as needed, reference their Ids [here](https://github.com/aplum/electron-builder/blob/4090ab62f16f05c5aa2bf7b2491415efd4594158/packages/app-builder-lib/src/targets/MsiTarget.ts#L266)), but I can't justify the time to figure out getting/converting the icon file since I don't need this feature at my job.

Also, I didn't use the FileAssociation.name config. I'm not sure what the expected values look like here, but the given example ("PNG") doesn't match up with the recommended format where it's used in [NsisTarget](https://github.com/electron-userland/electron-builder/blob/82ce942069dd54094951f6db890d335366048d4f/packages/app-builder-lib/src/targets/nsis/NsisTarget.ts#L685) – in the [docs](https://nsis.sourceforge.io/FileAssoc) the example value is "myapp.textfile", which matches the format used in WiX [ProgId Id examples](https://www.firegiant.com/wix/tutorial/getting-started/beyond-files/). So in NSIS it looks like we should generate that value ourselves, as done in MSI ProgId here. Furthermore, on [Mac targets](https://github.com/electron-userland/electron-builder/blob/96d7813362d319c3141d78e179c3f6d71c798fab/packages/app-builder-lib/src/electron/electronMac.ts#L203) `name` is used for CFBundleTypeName, which I'm not familiar with but [this example](https://developer.apple.com/library/archive/qa/qa1959/_index.html) uses "Adobe PDF" and iMovie.app on my Mac uses values like "Color Preset"… so I'm thinking maybe the `name` and `description` options should be merged?

Closes #3694, #6488

mmaietta commented:
Since we're targeting the v23 alpha branch, perhaps we can remove the iconId field and generate it ourselves as you mentioned? That way we can avoid this:
>it's now too specific – we also need a product-specific Id prefix for ProgId Ids with this PR, and possibly others in the future.

Not sure regarding the other items, the subject kind of flew over my head. Never worked with File Associations in the past

aplum commented:
I've removed the iconId config option and added a better fallback value than the "ElectronApp" I was using.

Supporting FileAssociation.icon could be done in a minor release in the future, and my rambling paragraph about FileAssociation.name would be more appropriate as a separate PR anyway (if it holds any merit) (FYI don't expect a PR about it from me), so I'm marking this as ready for review now.

mmaietta commented:
Can you please merge in latest master? That should enable latest test configuration to be run

aplum commented:
Done, I think?

mmaietta commented:
Not sure what happened, but the merge triggered much more "file changes" than what was originally in the PR.

Could you try rebasing your feature on top of `v23.0.0-alpha` branch? I just want to clean up the PR to only have your changes.

aplum commented:
That seems to have fixed it.

mmaietta commented:
Published in `23.0.0-alpha.0`. Enjoy!
